File: vehiculos-frontend/src/app/services/vehiculo.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Vehiculo {
  id: number;
  vin: string;
  placa: string;
  modelo: string;
  estatus: string;
}

@Injectable({
  providedIn: 'root'
})
export class VehiculoService {
  private apiUrl = 'http://localhost:8080/api/vehiculos'; // Ajustar al puerto del backend

  constructor(private http: HttpClient) {}

  // Obtener todos los vehículos
  getVehiculos(): Observable<Vehiculo[]> {
    return this.http.get<Vehiculo[]>(`${this.apiUrl}/consultar`);
  }

  // Obtener un vehículo por ID
  getVehiculo(id: number): Observable<Vehiculo> {
    return this.http.get<Vehiculo>(`${this.apiUrl}/consultar/${id}`);
  }

  // Insertar un nuevo vehículo
  insertarVehiculo(vehiculo: Vehiculo): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/insertar`, vehiculo);
  }

  // Editar un vehículo
  editarVehiculo(id: number, vehiculo: Vehiculo): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/editar/${id}`, vehiculo);
  }

  // Eliminar un vehículo
  eliminarVehiculo(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/eliminar/${id}`);
  }
}

File: vehiculos_backend/src/main/java/com/vehiculos/vehiculos_backend/services/VehiculoService.java
package com.vehiculos.vehiculos_backend.services;

import com.vehiculos.vehiculos_backend.models.Vehiculo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class VehiculoService {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    // Insertar un vehículo
    public void insertarVehiculo(String vin, String placa, String modelo, String estatus) {
        String sql = "{CALL insertar_vehiculo(?, ?, ?, ?)}";
        jdbcTemplate.update(sql, vin, placa, modelo, estatus);
    }

    // Editar un vehículo
    public void editarVehiculo(int id, String vin, String placa, String modelo, String estatus) {
        String sql = "{CALL editar_vehiculo(?, ?, ?, ?, ?)}";
        jdbcTemplate.update(sql, id, vin, placa, modelo, estatus);
    }

    // Consultar un vehículo
    public Vehiculo consultarVehiculo(int id) {
        String sql = "{CALL consultar_vehiculo(?)}";
        return jdbcTemplate.queryForObject(
                sql,
                new Object[]{id},
                (rs, rowNum) -> {
                    Vehiculo vehiculo = new Vehiculo();
                    vehiculo.setId(rs.getInt("ID"));
                    vehiculo.setVin(rs.getString("VIN"));
                    vehiculo.setPlaca(rs.getString("placa"));
                    vehiculo.setModelo(rs.getString("modelo"));
                    vehiculo.setEstatus(rs.getString("estatus"));
                    return vehiculo;
                });
    }


    // Consultar todos los vehículos
    public List<Vehiculo> consultarVehiculos() {
        String sql = "{CALL consultar_vehiculos()}";
        return jdbcTemplate.query(
                sql,
                (rs, rowNum) -> {
                    Vehiculo vehiculo = new Vehiculo();
                    vehiculo.setId(rs.getInt("ID"));
                    vehiculo.setVin(rs.getString("VIN"));
                    vehiculo.setPlaca(rs.getString("placa"));
                    vehiculo.setModelo(rs.getString("modelo"));
                    vehiculo.setEstatus(rs.getString("estatus"));
                    return vehiculo;
                }
        );
    }


    // Eliminar un vehículo
    public void eliminarVehiculo(int id) {
        String sql = "{CALL eliminar_vehiculo(?)}";
        jdbcTemplate.update(sql, id);
    }

}
